feat(rust): right-align numeric columns

alicja-januszkiewicz opened the pull request:
closes https://github.com/pola-rs/polars/issues/7378

alexander-beedie commented:
Hey, thanks for this  :) I put a follow-up comment on the original issue; do you think you could extend the PR to include an option for fixing the number of floating point decimal places offered? I suspect right-alignment would be more attractive/useful then.

Also, if you could make this opt-in for now, via a suitable `Config` option and associated environment variable, we can test it out a bit before deciding to change the default behaviour 👍

alicja-januszkiewicz commented:
Hey, thanks for the reply. This looks doable for me, I will give it a shot.

alexander-beedie commented:
> Hey, thanks for the reply. This looks doable for me, I will give it a shot.

Fantastic :) 
I'd suggest introducing two new config options via the following environment vars:

* `POLARS_FMT_DECIMAL_PLACES`, defaulting to None, otherwise a +ve integer.
* `POLARS_FMT_TABLE_CELL_NUMERIC_ALIGNMENT`, defaulting to None, allowing "LEFT", "RIGHT", "CENTER" (same as`POLARS_FMT_TABLE_CELL_ALIGNMENT`, which it should override for numeric cols, if not None).

Add them into the `Config` class on the python side (can pattern after `set_tbl_cell_alignment`), and then we can experiment with the two settings independently to see what might work best for a potential new default 👍

alexander-beedie commented:
Nice, looks like it's almost there - and `POLARS_FMT_FLOAT_PRECISION ` is a good choice (fits well with the existing `float_precision` param in write_excel/write_csv) 👍

alicja-januszkiewicz commented:
ready for review @alexander-beedie :)

alexander-beedie commented:
> ready for review @alexander-beedie :)

Will take a look tonight!

alexander-beedie commented:
FYI, I added a mechanism for allowing directly applied values (those not mediated by env vars) to better participate in `Config` load/save yesterday: https://github.com/pola-rs/polars/pull/7696.

alicja-januszkiewicz commented:
Right I've updated the PR to use that mechanism for float_precision.

This is sort of ready for review again, but there is a bit of an issue regarding the float_precision option. Namely, that when using `set_float_format` & `get_float_format` instead of using env vars, there isn't a straightforward way of checking if the option was ever set in the first place, so I have reserved the value of 255 (u8:MAX) to indicate an unset value. I admit this is a bit lazy but the alternative would require something like `FLOAT_PRECISION: RwLock<Option<AtomicU8>>`, which seems a bit over-complicated. Or perhaps that would have been better? What do you think?

alexander-beedie commented:
> I admit this is a bit lazy but the alternative would require something like `FLOAT_PRECISION: RwLock<Option<AtomicU8>>`, which seems a bit over-complicated. Or perhaps that would have been better? What do you think?

I think that's fine; keeps it simple - can you raise an error if anyone tries to set a nonsense value? Probably anything above 16, which (if I recall correctly) is about the precision limit of an `f64`. Certainly anything much above that is living in denial about floating point accuracy :)

Have had family in town, so got a bit backlogged; will download the latest updates at thew/e and run through them! 👍

alicja-januszkiewicz commented:
Sorry I was a bit busy myself over the Easter. I've tried to rebase tonight but I can't figure out how to without adding in hundreds of commits to the PR. I will try to sort this out next weekend.

alicja-januszkiewicz commented:
@alexander-beedie Hey sorry I only just got around to fixing this, I've added the check for float precision being greater than 16 and rebased.

> A rebase + force-push should do the trick?
Yeah that's what I tried originally, but I ended up getting some commits from main in the commits/changes tabs to this PR. It was probably because I merged main into this branch at some earlier point, so I tried removing that merge commit first and that seems to have fixed it.

stevenlis commented:
> I'd suggest introducing two new config options via the following environment vars:
> 
> * `POLARS_FMT_DECIMAL_PLACES`: default to None, otherwise a +ve integer.
> * `POLARS_FMT_TABLE_CELL_NUMERIC_ALIGNMENT`: default to None, allowing "LEFT", "RIGHT", "CENTER" - same as`POLARS_FMT_TABLE_CELL_ALIGNMENT`, which it should override for numeric cols (if not None).
> 
> Add them into the `Config` class on the python side (can pattern after `set_tbl_cell_alignment`), and then we can experiment with the two settings independently to see what might work best for a potential new default 👍

Sorry to bother you guys. But is this still happening? @alexander-beedie @alicja-januszkiewicz

alexander-beedie commented:
> Sorry to bother you guys. But is this still happening? @alexander-beedie @alicja-januszkiewicz
 
Argh... this is 100% my fault; I must have missed some GitHub notifications and let it slip through the cracks. Short version is that "yes", I'd love to get this merged - it's a great contribution, and (potentially) could even become the default format.

@alicja-januszkiewicz; since this is my bad, I have checked out this branch rebased/merged locally; unfortunately I don't seem able to push it back out, though the  _"Maintainers are allowed to edit this pull request"_ option has been checked. I'll see if I can work out why not and will update accordingly 🤔 

```     
error: failed to push some refs to 'https://github.com/alicja-januszkiewicz/polars.git'
To https://github.com/alicja-januszkiewicz/polars.git
!	refs/heads/right_aligned_numbers:refs/heads/right_aligned_numbers	[remote rejected] 
```

ritchie46 commented:
@alexander-beedie do you use the github cli? This always works much, much easier for pushing to a PR.

alexander-beedie commented:
> @alexander-beedie do you use the github cli?

I do not, but I'm willing to try... ;)

alicja-januszkiewicz commented:
Hi,

I'd be happy to continue working on this, but I'm not sure what is the next step here - this is still pending review right? @alexander-beedie

alexander-beedie commented:
> I'd be happy to continue working on this, but I'm not sure what is the next step here - this is still pending review right? @alexander-beedie

From my perspective I think it's in good shape; @ritchie46 or @orlp, would one of you mind having a quick look? (Note that this is a new formatting.alignment option, and not enabled by default). Then let's get it merged and we can iterate if necessary :)

orlp commented on crates/polars-core/src/fmt.rs:
```diff
@@ -33,6 +33,7 @@ pub enum FloatFmt {
     Full,
 }
 static FLOAT_FMT: AtomicU8 = AtomicU8::new(FloatFmt::Mixed as u8);
+static FLOAT_PRECISION: AtomicU8 = AtomicU8::new(u8::MAX);
```
Why is this an `u8`? There are floating point values that require more than 255 digits, e.g. `2^-1000` is exactly representable by an `f64` and requires 1000 decimal digits to print its full value.

alexander-beedie replied:
Digits vs precision; nothing here limits the total number of digits.

alexander-beedie replied:
I suppose we could make it arbitrary, but usefully there are only 16 digits of available precision in an `f64`, hence `u8`. I would be surprised if many people were hoping to print 1000 digit values, but I guess if somebody wants to and they have the terminal space, why stop them ;)

orlp commented on crates/polars-core/src/fmt.rs:
```diff
@@ -656,19 +665,24 @@ impl Display for DataFrame {
+                for (column_index, column) in table.column_iter_mut().enumerate() {
+                    let dtype = fields[column_index].data_type();
+                    let mut preset = str_preset.as_str();
+                    if dtype.to_physical().is_numeric() {
```
I don't think we should test if the *physical* type is numeric. Unless I'm mistaken this will right-align things such as categoricals and dates, and I don't think that was part of the original scope.

alexander-beedie replied:
Ahh, good catch 👍

orlp commented on crates/polars-core/src/fmt.rs:
```diff
@@ -710,6 +724,15 @@ const SCIENTIFIC_BOUND: f64 = 999999.0;
     let v: f64 = NumCast::from(v).unwrap();
+
+    let precision = get_float_precision();
+    if precision != u8::MAX {
```
Instead of using`u8::MAX` to signal 'no option set' I think we should use an `Option<T>` initialized with `None`.

alexander-beedie replied:
@alicja-januszkiewicz: want to take this one? :)  I'm going to take a look at the max precision value again, as we should consider Decimal too (support for which is going to be turned on by default soon).

alicja-januszkiewicz replied:
Sure, I'll try to get this over by the end of the week.

alicja-januszkiewicz replied:
I've changed the type of `FLOAT_PRECISION` from `AtomicU8` to `RwLock<Option<usize>>`.

I've picked `usize` for now as that's the type the `write!` and `format!` macros expect, so this saves having to repeatedly cast the value.

I've wrapped it in an `RwLock` to hopefully retain whatever protection `AtomicU8` was providing, as there doesn't seem to be an atomic equivalent of `Option`.

orlp commented on py-polars/polars/config.py:
```diff
@@ -314,6 +328,26 @@ def set_auto_structify(cls, active: bool = False) -> type[Config]:
+        if precision is None:
+            precision = 255
+        elif precision > 16 and precision != 255:
+            raise ValueError(
```
Why are we arbitrarily limiting precision to 16 digits?

orlp commented on py-polars/src/functions/meta.rs:
```diff
@@ -56,3 +56,21 @@ pub fn get_float_fmt() -> PyResult<String> {
+#[pyfunction]
+pub fn set_float_precision(precision: u8) -> PyResult<()> {
+    use polars_core::fmt::set_float_precision;
+    if precision > 16 && precision != u8::MAX {
```
Again, why is the maximum supported precision 16 digits?

orlp replied:
@alexander-beedie That really depends on what you're doing. Every floating point number has an *exact* decimal representation (the converse isn't true), and someone might care about that. Most often people just go for "round-trip correctness", but this leads to silly things as this

```rust
println!("{}", 1u64 << 63);
println!("{}", (1u64 << 63) as f64);
```

printing

```
9223372036854775808
9223372036854776000
```

Note that `9223372036854775808` *is* the exact value. The overemphasis to round-trip correctness has also lead to real-world data loss. Rust (and `serde`) by default
print `f32`s with less precision because "only a few digits is needed for a correct-roundtrip". The problem is that reading back the values as `f64` gives you *different* values than if you were to cast `f32` as `f64` (which is completely lossless).

So if you print `f64` values with only as few digits as needed to round-trip it is correct *if read back in as an f64*, but if you were to read it out as, say, `Decimal`, you would get **wrong values**.

```python
>>> Decimal(math.pi) == Decimal(str(math.pi))
False
```

orlp replied:
You may want to read this part of a blog post I made that goes into more detail: https://orlp.net/blog/ordering-numbers/#floating-point-numbers-are-exact.

svaningelgem commented:
@stinodego : This is a PR that I think could be merged as @alicja-januszkiewicz is still active on it (congratz mate!) and I think the feature would be well received.
CC: @alexander-beedie

alexander-beedie commented:
> @stinodego : This is a PR that I think could be merged as @alicja-januszkiewicz is still active on it (congratz mate!) and I think the feature would be well received.
> 
> CC: @alexander-beedie 

Give me a few mins on this one; has been a (too) long road getting this one over the line (my bad 😅)
